// edgegptjs/lib/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseProxyString = exports.getRanHex = exports.appendIdentifier = void 0;
const constants_1 = require("./constants");
const crypto_1 = require("crypto");
/**
 * Appends special character to end of message to identify end of message
 * @param msg
 * @returns
 */
function appendIdentifier(msg) {
    return JSON.stringify(msg) + constants_1.DELIMITER;
}
exports.appendIdentifier = appendIdentifier;
/**
 * Returns random hex string
 * @param length
 */
function getRanHex(length = 32) {
    return (0, crypto_1.randomBytes)(Math.ceil(length / 2))
        .toString('hex')
        .slice(0, length);
}
exports.getRanHex = getRanHex;
function parseProxyString(proxyString) {
    const url = new URL(proxyString);
    const result = {
        host: url.hostname,
        port: parseInt(url.port, 10),
    };
    if (url.protocol.startsWith('http') || url.protocol.startsWith('https')) {
        result.protocol = url.protocol;
    }
    if (url.username || url.password) {
        result.auth = {
            username: decodeURIComponent(url.username),
            password: decodeURIComponent(url.password),
        };
    }
    return result;
}
exports.parseProxyString = parseProxyString;
