// edgegptjs/lib/chathub/ChatBot.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncDelegator = (this && this.__asyncDelegator) || function (o) {
    var i, p;
    return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
    function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
};
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
    function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Chatbot = void 0;
const ChatHub_1 = require("./ChatHub"); // Import the ChatHub class from the previous code block
const Conversation_1 = require("./Conversation"); // Import the Conversation class
class Chatbot {
    constructor(cookiePath = '', cookies = null, proxy = null) {
        this.cookiePath = cookiePath;
        this.cookies = cookies;
        this.proxy = proxy;
        this.chatHub = null;
        this.cookiePath = cookiePath;
        this.cookies = cookies;
        this.proxy = proxy;
        const conversation = new Conversation_1.Conversation(cookiePath, cookies, proxy);
        this.chatHubInitialization = conversation.initialize();
        this.chatHubInitialization.then(() => (this.chatHub = new ChatHub_1.ChatHub(conversation, proxy)));
    }
    ask(prompt, wssLink = 'wss://sydney.bing.com/sydney/ChatHub', conversationStyle = null) {
        var _a, e_1, _b, _c;
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.chatHub) {
                throw new Error('the chathub has not initialized!');
            }
            try {
                for (var _d = true, _e = __asyncValues(this.chatHub.askStream(prompt, wssLink, conversationStyle)), _f; _f = yield _e.next(), _a = _f.done, !_a;) {
                    _c = _f.value;
                    _d = false;
                    try {
                        const [final, response] = _c;
                        if (final) {
                            return response;
                        }
                    }
                    finally {
                        _d = true;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = _e.return)) yield _b.call(_e);
                }
                finally { if (e_1) throw e_1.error; }
            }
            if (this.chatHub.wss) {
                this.chatHub.wss.close();
            }
            return null;
        });
    }
    askStream(prompt, wssLink = 'wss://sydney.bing.com/sydney/ChatHub', conversationStyle = null) {
        return __asyncGenerator(this, arguments, function* askStream_1() {
            if (!this.chatHub) {
                throw new Error('the chathub has not initialized!');
            }
            yield __await(yield* __asyncDelegator(__asyncValues(this.chatHub.askStream(prompt, wssLink, conversationStyle))));
        });
    }
    close() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.chatHub) {
                throw new Error('the chathub has not initialized!');
            }
            yield this.chatHub.close();
        });
    }
    reset() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.close();
            this.chatHub = new ChatHub_1.ChatHub(new Conversation_1.Conversation(this.cookiePath, this.cookies), this.proxy);
        });
    }
}
exports.Chatbot = Chatbot;
